hack/generator/pkg/astmodel/struct_type.go

```diff
@@ -110,14 +109,42 @@ func (structType *StructType) Equals(t Type) bool {
 	return false
 }
 
-// CreateRelatedDefinitions implements the HasRelatedDefinitions interface for StructType
-func (structType *StructType) CreateRelatedDefinitions(ref PackageReference, namehint string, idFactory IdentifierFactory) []Definition {
-	var result []Definition
-	for _, f := range structType.fields {
-		nh := namehint + "." + string(f.fieldName)
-		defns := f.CreateRelatedDefinitions(ref, nh, idFactory)
-		result = append(result, defns...)
+func (st *StructType) CreateDefinitions(name *TypeName, idFactory IdentifierFactory, isResource bool) (TypeDefiner, []TypeDefiner) {
+
+	ref := NewStructReference(name.name, name.groupName, name.packageName, isResource)
+
+	var otherTypes []TypeDefiner
+	var newFields []*FieldDefinition
+
+	for _, field := range st.fields {
+		newField := field
+
+		// TODO: figure out a generic way to do this:
+		if et, ok := newField.FieldType().(*EnumType); ok {
```

Comment:
💥 
This will only work for enums directly used by fields. If the enum is wrapped by `MapType`, `ArrayType` or `OptionalType` it won't get a name. The previous implementation used recursion to handle this - and it broke when `OptionalType` was introduced.

Reply:
right, that's the Outstanding bit that I need to reintroduce